lib/WebAssemblyParser.js

```diff
@@ -17,37 +60,41 @@ class WebAssemblyParser extends Tapable {
 		this.options = options;
 	}
 
-	parse(source, state, callback) {
-		// TODO parse WASM AST and walk it
-		// TODO extract imports
-
+	parse(binary, state) {
 		// flag it as ESM
 		state.module.buildMeta.exportsType = "namespace";
 
-		// extract exports
-		// TODO find more efficient way doing it
-		// TODO use Promises
-		if (typeof WebAssembly !== "undefined") {
-			WebAssembly.compile(source)
-				.then(module => {
-					state.module.buildMeta.providedExports = WebAssembly.Module.exports(
-						module
-					).map(exp => exp.name);
-					for (const imp of WebAssembly.Module.imports(module)) {
-						const dep = new WebAssemblyImportDependency(
-							imp.module,
-							imp.name,
-							imp.kind
-						);
-						state.module.addDependency(dep);
-					}
-				})
-				.then(() => callback(null, state), err => callback(err));
-		} else {
-			throw new Error(
-				"Can't compile WebAssembly modules without WebAssembly support in current node.js version (Update to latest node.js version)"
-			);
-		}
+		// parse it
+		const ast = decode(binary, decoderOpts);
+
+		// extract imports and exports
+		const exports = (state.module.buildMeta.providedExports = []);
+		traverse(ast, {
+			ModuleExport({ node }) {
+				exports.push(node.name);
+			},
+
+			ModuleImport({ node }) {
+				const dep = new WebAssemblyImportDependency(
+					node.module,
+					node.name,
+					node.descr
+				);
+
+				state.module.addDependency(dep);
+
+				// To avoid unexpected behaviors we check that the Memory or Table
+				// has been exported and is not accessed by the user.
```

Comment:
I would like to warn the error when we have unsupported stuff. This is of course not the proper way of doing it but it's just an idea for now. What do you think?

Could you please indicate me how I could parse and traverse the module?